server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java

```diff
@@ -205,6 +205,8 @@ protected void doInternalExecute(
         ActionListener<BulkResponse> listener,
         long relativeStartTimeNanos
     ) {
+        assert (bulkRequest instanceof SimulateBulkRequest) == false
+            : "TransportBulkAction should never be called with a SimulateBulkRequest";
```

Comment:
Should we also add an assert that there are no template substitutions here? I.e., that `getComponentTemplateSubstitutions()` returns an empty map?